authenticate on every request to broker when needed

flucivja opened the pull request:
Fixes #1240

Nevon commented:
Is this issue really limited to heartbeats? From reading the issue description, it sounds like the problem should occur on any request if the broker needs to reauthenticate.

If that's the case, then we could do the same thing in [sendRequest](https://github.com/tulios/kafkajs/blob/f443fc5f61fa09110bde7e9cee55c96d358a52ad/src/broker/index.js#L916), except when the request is `apiVersions`, since that happens before authentication.

Nevon commented on src/broker/index.js:
```diff
@@ -103,35 +110,45 @@ module.exports = class Broker {
+   * @private
+   * @returns {Promise}
+   */
+  async authenticate() {
```
If this can be called from other places than the connect method, then this also needs to also be within a lock so that you don't get multiple authentication processes going on simultaneously towards the same broker.

flucivja replied:
good point, I will add it

flucivja commented:
> Is this issue really limited to heartbeats? From reading the issue description, it sounds like the problem should occur on any request if the broker needs to reauthenticate.
> 
> If that's the case, then we could do the same thing in [sendRequest](https://github.com/tulios/kafkajs/blob/f443fc5f61fa09110bde7e9cee55c96d358a52ad/src/broker/index.js#L916), except when the request is `apiVersions`, since that happens before authentication.

yes, it can happend at any point. I will add it into [sendRequest]. 
btw. how to check if request is not `apiVersions`? By simply comparing apiName with string? Or better to get version protocol by `requests.ApiVersions.protocol` and compare names?

Nevon commented:
> btw. how to check if request is not apiVersions? By simply comparing apiName with string? Or better to get version protocol by requests.ApiVersions.protocol and compare names?

Not sure it matters much, but you could grab the api key of ApiVersions from `src/protocol/requests/apiKeys.js` and compare it with `request.apiKey`. The `apiName` shouldn't change, but `apiKey` definitely won't.

Nevon commented on src/broker/index.js:
```diff
@@ -63,25 +73,37 @@ module.exports = class Broker {
     // The lock timeout has twice the connectionTimeout because the same timeout is used
     // for the first apiVersions call
-    const lockTimeout = 2 * this.connection.connectionTimeout + this.authenticationTimeout
+    const lockTimeout = 2 * this.connection.connectionTimeout
```
This still needs to have the `authenticationTimeout` added, since we're still authenticating within the same flow.

flucivja commented:
@Nevon everything should be ok now. Can you please check? :) Thanks.

Nevon commented on src/broker/index.js:
```diff
@@ -71,17 +81,29 @@ module.exports = class Broker {
     })
 
+    this.authLock = new Lock({
+      timeout: this.authenticationTimeout,
```
Let's just add a little 50ms buffer to the lock timeout. The reason being that if there is an authentication timeout, we want the error to come from the authentication timeout and not a generic lock timeout. The lock timeout is just there to ensure that we don't hold a lock forever.

flucivja replied:
good thinking, I'll add it

flucivja replied:
done

Nevon commented:
I've been trying to figure out how to test this, and I wrote a test that looks something like this:

```js
test('Reauthenticates if needed before making a request', async () => {
  const saslEntry = saslEntries.pop()
  const connection = createConnection(saslEntry.opts())
  broker = new Broker({
    connection,
    logger: newLogger(),
  })
  await broker.connect()
  const spy = jest.spyOn(connection, 'send').mockImplementation()

  broker.authenticatedAt = null
  await broker.listGroups()

  // Assert that sasl authenticate requests happen
  console.log(spy.mock.calls)
})
```

But I ran into the problem that the server rejects the second authentication request with an ILLEGAL_SASL_STATE error. This got me thinking and reading the [KIP about reauthentication](https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate).

In the current implementation, if you have multiple requests going on when you realize that you have to reauthenticate, each one is going to try to reauthenticate, which is gonna cause a problem. This is relatively easily solved by having the `isAuthenticated` method wait for the `authLock` to be released, so that you wait to check if you need to authenticate while there's an authentication going on. Unfortunately it makes the method async, which ripples through and affects a lot of things. A pain in the ass to fix, but doable.

flucivja commented:
What do you think about putting `isAuthenticated` check into `[PRIVATE.AUTHENTICATE]()` after auth lock? And if it will be authenticated, then it will skip auth and release lock.

t-d-d commented:
You might want to look at the sharedPromise in utils.  Heartbeat() is
wrapped with this so that when multiple are issues concurrently, only 1
actual request is sent.

On Tue, 30 Nov 2021, 17:21 Tommy Brunn, ***@***.***> wrote:

> I've been trying to figure out how to test this, and I wrote a test that
> looks something like this:
>
> test('Reauthenticates if needed before making a request', async () => {
>   const saslEntry = saslEntries.pop()
>   const connection = createConnection(saslEntry.opts())
>   broker = new Broker({
>     connection,
>     logger: newLogger(),
>   })
>   await broker.connect()
>   const spy = jest.spyOn(connection, 'send').mockImplementation()
>
>   broker.authenticatedAt = null
>   await broker.listGroups()
>
>   // Assert that sasl authenticate requests happen
>   console.log(spy.mock.calls)})
>
> But I ran into the problem that the server rejects the second
> authentication request with an ILLEGAL_SASL_STATE error. This got me
> thinking and reading the KIP about reauthentication
> <https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate>
> .
>
> In the current implementation, if you have multiple requests going on when
> you realize that you have to reauthenticate, each one is going to try to
> reauthenticate, which is gonna cause a problem. This is relatively easily
> solved by having the isAuthenticated method wait for the authLock to be
> released, so that you wait to check if you need to authenticate while
> there's an authentication going on. Unfortunately it makes the method
> async, which ripples through and affects a lot of things. A pain in the ass
> to fix, but doable.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tulios/kafkajs/pull/1241#issuecomment-982440296>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDLW5TG5ZGLFA6B4MGVTFTUOSJQLANCNFSM5I6ZIHQQ>
> .
>

flucivja commented:
`sharedPromise` looks handy for such case. But I am thinking if this is really an issue. I tried that test without my changes and it is failing too. @Nevon is that test passing for you without my changes (without check for re-authentication)?

Nevon commented:
> @Nevon is that test passing for you without my changes (without check for re-authentication)?

No, but I think it's an issue with the test. Maybe you can't authenticate again right after authenticating with sasl/scram like I'm doing. I'm guessing if you swap it out for oauthbearer with a short reauthentication timeout maybe it'll work. I just didn't have time to test it out.

> But I am thinking if this is really an issue.

I would be very surprised if it isn't. Tossing `broker.listGroups` or something into a `Promise.all` should trigger it immediately.

flucivja commented:
Ok then, are you ok to handle it with `sharedPromise`? It will look like this in the broker constructor:

```
this[PRIVATE.AUTHENTICATE] = sharedPromiseTo(async () => {
  ... authentication logic
})
```

Nevon commented:
Sounds like a good solution! Let's give it a try. I would also really like to test these scenarios. We do have a test suite that uses oauthbearer, which should certainly allow for reauthentication. See the `test:group:oauthbearer` yarn script, which runs a subset of test files. There's a function called `describeIfOauthbearerEnabled` in `testHelpers` that will run tests nested in it only if oauthbearer is enabled.

flucivja commented:
Hi @Nevon, I implemented tests and sorry it took longer time. I had issues with kafka SASL OAUTHBEARER authentication. Within test, the kafka returned me all the time that provided token created in test utils is invalid and I didn't know why. After debugging and looking to how other kafka clients are implemeneted, I found out that [expiration time](https://github.com/tulios/kafkajs/pull/1241/files#diff-889932f4cee3b60450073a0c903c4e885fb045142b8343a6ade91c4020b5419a) was missing in token.

Btw. you were right with multiple pararel requests causing error: when I removed fix with `sharedPromise` my tests which runs multuple requests in pararel requiring authentication started failing with invalid SASL state.

Can you please check PR again when you will have time? If all will be good, can you merge it and release?

Thank you very much.

flucivja commented:
@Nevon just a gentle reminder :)

Nevon commented on src/broker/__tests__/reauthenticate.spec.js:
```diff
@@ -0,0 +1,118 @@
+const Broker = require('../index')
+
+/**
+ * Session lifetime on broker is configured in docker-compose.2_4_oauthbearer.yml with 15000ms in env KAFKA_CONNECTIONS_MAX_REAUTH_MS
```
Is it possible to reauthenticate much sooner than this? I'm thinking maybe we can avoid the excessive sleeps in the test by either setting the `reauthenticationThreshold` to like 14900ms or something so we only need to sleep a little bit, or override `broker.sessionLifetime` to be 0 whenever we want to trigger a reauthentication. Something like:

```ts
    const connection = createConnection(saslOAuthBearerConnectionOpts())
    const spy = jest.spyOn(connection, 'send')
    broker = new Broker({
      connection,
      logger: newLogger(),
    })
    await broker.connect()
    await broker.listGroups()
    await broker.listGroups()

    broker.sessionLifetime = 0n;
    await broker.listGroups()

    expect(getAuthCalls(spy).length).toBe(2)
```

With all the sleeps in the current test, we're adding about half a minute or more to the execution time that we can maybe avoid.

flucivja replied:
yes sure, I will reduce it by using higher `reauthenticationThreshold `.

Nevon commented:
Haven't forgotten about you. I left the notification sitting on my phone so it wouldn't get lost. 😅  The change looks good overall - I'd just like to get your thoughts on whether we can get rid of those sleeps in the test (or at least reduce them).

Nevon commented on src/broker/index.js:
```diff
@@ -62,16 +73,55 @@ module.exports = class Broker {
+     */
+    this[PRIVATE.AUTHENTICATE] = sharedPromiseTo(async () => {
+      try {
+        await this.authLock.acquire()
```
We don't need this lock anymore now right? If there's more callers, the first one will create a promise and start the authentication, the second one will just receive the promise from the first one. Since the promise creation itself is synchronous there shouldn't be any race condition there, I think.

flucivja replied:
yes you are right. I will remove auth lock.

flucivja commented:
@Nevon done, now my test is faster by 30 seconds (14s vs 44s).

Nevon commented:
Thanks for your patience and for taking the time to contribute!

j-a-h-i-r commented:
Sorry for reviving a merged thread but I'm curious about one thing @Nevon. Does this fix https://github.com/tulios/kafkajs/issues/1039? As far as I can remember, that issue occurred due to re-authentication. I saw that you mentioned testing SASL OauthBearer flows. So I was wondering if the changes potentially fix that issue too.
